library/spdm_common_lib/libspdm_com_opaque_data.c

```diff
@@ -392,51 +492,42 @@ libspdm_process_opaque_data_version_selection_data(libspdm_context_t *spdm_conte
-    *spdm_general_opaque_data_table_header;
-    secured_message_opaque_element_table_header_t
+    const secured_message_opaque_element_table_header_t
     *opaque_element_table_header;
-    secured_message_opaque_element_version_selection_t
+    const secured_message_opaque_element_version_selection_t
     *opaque_element_version_section;
+
+    bool result;
     uint8_t secured_message_version_index;
+    const void *get_element_ptr;
+
+    result = false;
 
     if (spdm_context->local_context.secured_message_version
         .spdm_version_count == 0) {
         return LIBSPDM_STATUS_SUCCESS;
     }
 
-    if (data_in_size !=
-        libspdm_get_opaque_data_version_selection_data_size(spdm_context)) {
-        return LIBSPDM_STATUS_INVALID_MSG_SIZE;
-    }
-    if (libspdm_get_connection_version (spdm_context) >= SPDM_MESSAGE_VERSION_12) {
-        spdm_general_opaque_data_table_header = data_in;
-        if (spdm_general_opaque_data_table_header->total_elements != 1) {
-            return LIBSPDM_STATUS_INVALID_MSG_FIELD;
-        }
-        opaque_element_table_header =
-            (void *)(spdm_general_opaque_data_table_header + 1);
-    } else {
-        general_opaque_data_table_header = data_in;
-        if ((general_opaque_data_table_header->spec_id !=
-             SECURED_MESSAGE_OPAQUE_DATA_SPEC_ID) ||
-            (general_opaque_data_table_header->opaque_version !=
-             SECURED_MESSAGE_OPAQUE_VERSION) ||
-            (general_opaque_data_table_header->total_elements != 1)) {
-            return LIBSPDM_STATUS_INVALID_MSG_FIELD;
-        }
-        opaque_element_table_header =
-            (void *)(general_opaque_data_table_header + 1);
+    result = libspdm_get_element_from_opaque_data(
+        spdm_context, data_in_size,
+        data_in, SPDM_REGISTRY_ID_DMTF,
+        SECURED_MESSAGE_OPAQUE_ELEMENT_SMDATA_ID_VERSION_SELECTION,
+        &get_element_ptr);
+    if (!result) {
+        LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO,"get element error!\n"));
+        return LIBSPDM_STATUS_INVALID_MSG_FIELD;
     }
-    if ((opaque_element_table_header->id != SPDM_REGISTRY_ID_DMTF) ||
-        (opaque_element_table_header->vendor_len != 0) ||
+
+    opaque_element_table_header = (const secured_message_opaque_element_table_header_t*)
+                                  get_element_ptr;
+
+    /*check for selection vesion data*/
+    if ((opaque_element_table_header->vendor_len != 0) ||
         (opaque_element_table_header->opaque_element_data_len !=
          sizeof(secured_message_opaque_element_version_selection_t))) {
         return LIBSPDM_STATUS_INVALID_MSG_FIELD;
     }
-    opaque_element_version_section = (void *)(opaque_element_table_header + 1);
+    opaque_element_version_section = (const void *)(opaque_element_table_header + 1);
     if ((opaque_element_version_section->sm_data_version !=
          SECURED_MESSAGE_OPAQUE_ELEMENT_SMDATA_DATA_VERSION) ||
         (opaque_element_version_section->sm_data_id !=
```

Comment:
why we need check again?

I think we need move sm_data_version check to common function and remove check here.